### ISSUE TITLE
Distinguish types of calls for call hierarchy?

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#84100]
For the call hierarchy, @rbuckton brought up that JavaScript and TypeScript has many different ways of actually invoking a function:

- New `new Foo()`
- Tagged templates: foo\`bar\`
- Decorators `@foo`
- Property/Element access to accessors `obj.foo`

My proposal is that TS should return call hierarchy items for all of these cases, but I wanted to check this with you @jrieken. Does that make sense? Should we consider having a way to distinguish different call types?

**Comment by jrieken** [comment]
Yeah, I would say always return all calls but having something like a `CallHierarchyCallKind`-property on `CallHierarchyIncomingCall` and `CallHierarchyOutgoinhCall` sound reasonable. It will be nice if TS starts on this and then we adjust the UX for it

**Comment by mjbvz** [comment]
@rbuckton Are there use cases where it would help users to be able to distinguish these different types of calls?

**Comment by rbuckton** [comment]
I don't have any specific use cases in mind, but I've seen similar features in other tools (such as Reflector) respect this kind of classification. For accessors especially, it can be useful to distinguish between call sites using the setter and call sites using the getter. Many hierarchy items will only have a single classification (classes can only be `new`-ed, most functions will only be called, etc.).

**Comment by jrieken** [comment]
>  For accessors especially, it can be useful to distinguish between call sites using the setter and call sites using the getter. 

I like and I believe that it will add value, e.g. the UX can show icons or allow to filter/group by kinds. In fact, there is a similar request to classify reference search results like that too, see https://github.com/microsoft/vscode/issues/74237. 

@rbuckton Would you be OK to invent kinds that makes sense for TS and we take that as base to make this vscode API - basically a common denominator that works for multiple languages

**Comment by rbuckton** [comment]
Possibly, I'll see what I can come up with.

**Comment by rbuckton** [comment]
So far it doesn't seem like there's a good reason to distinguish these types using icons:

![image](https://user-images.githubusercontent.com/3902892/68803201-277f7b00-0614-11ea-9196-983551364357.png)

![image](https://user-images.githubusercontent.com/3902892/68803219-2e0df280-0614-11ea-9f13-0cadb2aa5c3d.png)

However, I could see a use-case for filtering the results, possibly in the upper-right-hand corner of the peek adjacent to the "Calls To"/"Calls From" icon:

![image](https://user-images.githubusercontent.com/3902892/68803383-7decb980-0614-11ea-9693-5a8b268b38e4.png)

VS's version of Call Hierarchy does not have filtering, so I am not certain if there would be value in adding that capability.
